src/Compilers/CSharp/Portable/Syntax/SyntaxNodeExtensions.cs

```diff
@@ -71,7 +71,11 @@ internal static bool CanHaveAssociatedLocalBinder(this CSharpSyntaxNode syntax)
                 kind == SyntaxKind.Attribute ||
                 kind == SyntaxKind.ArgumentList ||
                 kind == SyntaxKind.ArrowExpressionClause ||
-                (syntax is ExpressionSyntax && (syntax.Parent as LambdaExpressionSyntax)?.Body == syntax);
+                (syntax is ExpressionSyntax && 
+                    ((syntax.Parent as LambdaExpressionSyntax)?.Body == syntax ||
+                     (syntax.Parent as SwitchStatementSyntax)?.Expression == syntax ||
+                     (syntax.Parent as ForEachStatementSyntax)?.Expression == syntax ||
+                     (syntax.Parent as IfStatementSyntax)?.Condition == syntax));
```

Comment:
I'll add a comment instead. Because I do not believe it would be appropriate to share this function with semantic model because logically they are testing for different things even though today the things appear to be identical. I can easily see them diverging in the future.